lib/server.go

```diff
@@ -2026,25 +2163,52 @@ func (srv *Server) _handleGetTransactions(pp *Peer, msg *MsgDeSoGetTransactions)
 	pp.AddDeSoMessage(msg, true /*inbound*/)
 }
 
-func (srv *Server) ProcessSingleTxnWithChainLock(
-	pp *Peer, txn *MsgDeSoTxn) ([]*MempoolTx, error) {
+func (srv *Server) ProcessSingleTxnWithChainLock(pp *Peer, txn *MsgDeSoTxn) ([]*MsgDeSoTxn, error) {
 	// Lock the chain for reading so that transactions don't shift under our feet
 	// when processing this bundle. Not doing this could cause us to miss transactions
 	// erroneously.
 	//
 	// TODO(performance): We should probably do this less frequently.
 	srv.blockchain.ChainLock.RLock()
-	defer func() {
-		srv.blockchain.ChainLock.RUnlock()
-	}()
+	defer srv.blockchain.ChainLock.RUnlock()
+
 	// Note we set rateLimit=false because we have a global minimum txn fee that should
 	// prevent spam on its own.
-	return srv.mempool.ProcessTransaction(
-		txn, true /*allowUnconnectedTxn*/, false, /*rateLimit*/
-		pp.ID, true /*verifySignatures*/)
+
+	// Only attempt to add the transaction to the PoW mempool if we're on the
+	// PoW protocol. If we're on the PoW protocol, then we use the PoW mempool's
+	// txn validity checks to signal whether the txn has been added or not. The PoW
+	// mempool has stricter txn validity checks than the PoS mempool, so this works
+	// out conveniently, as it allows us to always add a txn to the PoS mempool.
+	blockHeight := srv.blockchain.blockTip().Height
+	if blockHeight < srv.blockchain.params.ForkHeights.ProofOfStake2ConsensusCutoverBlockHeight {
+		_, err := srv.mempool.ProcessTransaction(
+			txn,
+			true,  /*allowUnconnectedTxn*/
+			false, /*rateLimit*/
+			pp.ID,
+			true, /*verifySignatures*/
+		)
+
+		// If we're on the PoW chain, and the txn doesn't pass the PoW mempool's validity checks, then
+		// it's an invalid txn.
+		if err != nil {
+			return nil, errors.Wrapf(err, "Server.ProcessSingleTxnWithChainLock: Problem adding transaction to PoW mempool: ")
+		}
+	}
+
+	// Regardless of the consensus protocol we're running (PoW or PoS), we use the PoS mempool's to house all
+	// mempool txns. If a txn can't make it into the PoS mempool, which uses a looser unspent balance check for
+	// the the transactor, then it must be invalid.
+	if err := srv.posMempool.AddTransaction(NewMempoolTransaction(txn, uint64(time.Now().UnixMicro())), true); err != nil {
```

Comment:
This makes me a little nervous as there might be edge cases where it's rejected that we haven't thought of, but I can't think of any so probably OK.

Reply:
Agreed. If a txn is accepted by the PoW mempool, then it must be accepted by the PoS mempool too. The PoS mempool has a strictly relaxed set of txn acceptance rules compared to its PoW counterpart. I can't think of any edge cases either, so this *should* be safe